extension/packagemanagerextension/serverextension/envmanager.py

```diff
@@ -0,0 +1,329 @@
+        kerneljson = {
+            "argv": [js['prefix'] + "/bin/python",   "-m",
+                     "ipykernel_launcher",
+                     "-f",
+                     "{connection_file}"],
+            "display_name": "Python (" + folder + ")",
+            "language": "python"
+        }
+        kdir = '.local/share/jupyter/kernels/' + name
+        if not os.path.exists(kdir):
+            os.makedirs(kdir)
+        with open(kdir + '/kernel.json', 'w') as fp:
+            json.dump(kerneljson, fp)
+
+        data['PACKAGE_INFO'] = packages
+        with open(directory + '.swanproject', 'w') as outfile:
+            yaml.dump(data, outfile, default_flow_style=False)
+        return op
+
+    def env_packages(self, env):
+        output = self._execute('conda list --no-pip --json -n', env)
+        data = self.clean_conda_json(output)
+        if 'error' in data:
+            # we didn't get back a list of packages, we got a dictionary with
+            # error info
+            return data
+
+        return {
+            "packages": [pkg_info(package) for package in data]
+        }
+
+    def check_update(self, directory):
+        env = ""
+        directory = str(directory) + ".swanproject"
+        try:
+            env = yaml.load(open(directory))['ENV']
+        except:
+            return {'error': "Can't find project"}
+        packagesJson = self._execute('conda list --no-pip --json -n', env)
+        packagesJson = self.clean_conda_json(packagesJson)
+        packages = []
+        for it in packagesJson:
+            packages.append(it.get('name'))
+        output = self._execute('conda update --dry-run -q --json -n', env,
+                               *packages)
+        data = self.clean_conda_json(output)
+
+        if 'error' in data:
+            # we didn't get back a list of packages, we got a dictionary with
+            # error info
+            return data
+        elif 'actions' in data:
+            links = data['actions'].get('LINK', [])
+            package_versions = [link.get('dist_name') for link in links]
+            return {
+                "updates": [pkg_info(pkg_version)
+                            for pkg_version in package_versions]
+            }
+        else:
+            # no action plan returned means everything is already up to date
+            return {
+                "updates": []
+            }
+
+    def install_packages(self, directory, packages):
+        env = ""
+        directory = str(directory) + ".swanproject"
+        try:
+            env = yaml.load(open(directory))['ENV']
+        except:
+            return {'error': "Can't find project"}
+        output = self._execute('conda install -y -q --json -n', env, *packages)
+        data = {'ENV': env}
+        packages = self._execute('conda list --no-pip --json -n', env)
+        packages = self.clean_conda_json(packages)
+        data['PACKAGE_INFO'] = packages
+        with open(directory, 'w') as outfile:
```

Comment:
the exact same thing in writing to the swanproject. For this project I see some actions required in lot of places: reading swanproj, writing swanproj, reading packages from env, getting the name of the file swanproj.